### PR TITLE
improvement(build-tools): Separate fluid-build and flub configs

### PR DESCRIPTION
**Pull request by tylerbutler**
[AB#12470](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/12470)

Separates the fluid-build config settings - `tasks` and `repoPackages` - into its own config file and loads those settings separately from the rest of the flub settings. Eventually we want to fully separate the common fluid-build and repo infrastructure fully from the CLI - this change is a step towards that by formally separating the configs between the two tools.

The fluidBuild config stays in the existing file - so `tasks` and `repoPackages` continue to be defined in `fluidBuild.config.cjs`.

The other settings, including policy settings, have been moved to build-cli and loaded directly from a new config file, `flub.config.cjs`. The TypeScript interface is called `FlubConfig`.

For back-compat, the flub config loading path will load from `fluidBuild.config.cjs` unless the replacement flub-specific config file is present. This allows consumers to upgrade build-tools then separately split out their config settings into the new files. Eventually we can remove the loading of the old file in the flub conde path.

I also added a version field to both configs - `1` is the only supported value right now.

Other changes:

- Upgrades cosmiconfig to latest 8.x release.
- The FluidRepo construction pattern has changed to use the constructor directly rather than a factory method.
- The `Context` now exposes both the fluid build config and the flub config as separate properties. Calling code has been updated to use the appropriate config where needed.
- Splits the type test config into its own file for organization purposes. Config still lives in the build-tools project but should ultimately move along with all the type generation code into build-cli. That can be done separately, and is tracked by [AB#12469](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/12469).

### CODE REVIEW

File: build-tools/packages/build-cli/src/library/commands/base.ts [hunk 1]
```diff
@@ -137,19 +146,27 @@ export abstract class BaseCommand<T extends typeof Command>
...
+
+	// protected async getFlubConfig(configFile?: string): Promise<FlubConfig> {
+	// 	if (this._config === undefined) {
+	// 		const context = await this.getContext();
+	// 		this._config = getFlubConfig(configFile ?? context.repo.resolvedRoot);
+	// 	}
+
+	// 	return this._config;
+	// }
+
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/library/commands/base.ts:169]:
```suggestion
```

File: build-tools/packages/build-tools/src/common/fluidUtils.ts [hunk 2]
```diff
@@ -127,40 +127,36 @@ export async function getResolvedFluidRoot(buildRoot = false) {
 	return await realpathAsync(resolvedRoot);
 }
 
+const configName = "repoBuild";
```
Comment (curtisman) [inline build-tools/packages/build-tools/src/common/fluidUtils.ts:130]: Why introduce a new config name for "fluid-build"?

File: build-tools/packages/build-tools/src/common/typeValidatorConfig.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/*!
```
Comment (curtisman) [inline build-tools/packages/build-tools/src/common/typeValidatorConfig.ts:1]: Would be nice to move type validator out of fluid-build in the future.

**Review by curtisman — APPROVED** [review]

File: build-tools/packages/build-tools/src/common/fluidUtils.ts [hunk 4]
```diff
@@ -127,40 +127,36 @@ export async function getResolvedFluidRoot(buildRoot = false) {
...
-/**
- * Get an IFluidBuildConfig from the fluidBuild property in a package.json file, or from fluidBuild.config.[c]js.
+ * Get an IRepoBuildConfig from the fluidBuild property in a package.json file, or from fluidBuild.config.[c]js.
  *
  * @param rootDir - The path to the root package.json to load.
  * @param noCache - If true, the config cache will be cleared and the config will be reloaded.
  * @returns The fluidBuild section of the package.json, or undefined if not found
  */
-export function getFluidBuildConfig(rootDir: string, noCache = false): IFluidBuildConfig {
+export function getRepoBuildConfig(rootDir: string, noCache = false): IRepoBuildConfig {
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-tools/src/common/fluidUtils.ts:159]: Do you mean I should add a "version" field to the configs themselves? Or something else? I'm not quite sure what you mean by schema in this case.

File: build-tools/packages/build-tools/src/common/typeValidatorConfig.ts (hunk 3 above)
Comment (tylerbutler) [reply build-tools/packages/build-tools/src/common/typeValidatorConfig.ts:1]: I have work scheduled this sprint to do it.

File: build-tools/packages/build-tools/src/common/fluidUtils.ts (hunk 2 above)
Comment (tylerbutler) [reply build-tools/packages/build-tools/src/common/fluidUtils.ts:130]: I reverted these changes back to fluidBuild. If/when we have a new name we can make the changes then.

File: build-tools/packages/build-tools/src/common/fluidUtils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-tools/src/common/fluidUtils.ts]: Added a version field and checked it - if not set or set to unspported value (anything except `1`) throw an exception.

File: build-tools/packages/build-tools/src/common/fluidUtils.ts [hunk 5]
```diff
@@ -165,6 +151,18 @@ export function getFluidBuildConfig(rootDir: string, noCache = false): IFluidBui
...
-	return config?.config;
+	const configResult = configExplorer.search(rootDir);
+	const config = configResult?.config as IFluidBuildConfig | undefined;
+
+	if (config === undefined) {
+		throw new Error("No fluidBuild configuration found.");
+	}
+
+	// Only version 1 of the config is supported. If any other value is provided, throw an error.
+	if (config?.version !== FLUIDBUILD_CONFIG_VERSION) {
```
Comment (curtisman) [inline build-tools/packages/build-tools/src/common/fluidUtils.ts:162]: Should you allow version to be "undefined"?

Comment (tylerbutler) [reply build-tools/packages/build-tools/src/common/fluidUtils.ts:162]: Hmmm, I guess the "safe" thing to do is make it nullable and coerce it to a default value, but I'd like to make it required in a future release. For now I'll warn if it's undefined and error if it's an unsupported version.

**Pull request merged**